### PR TITLE
Remove winston-daily-rotate-file in favor of custom implementation

### PR DESCRIPTION
**Pull request by niik**
## Description

As noted in #14152 our attempts to remove `moment.js` got thwarted at the last minute when we realized that `winston-daily-rotate-file` depended on it.

This is an attempt to rip out `winston-daily-rotate-file` in favor of a custom file rotator.

N.B if this lands there's very little of winston's logic left so it's worth thinking about simplifying even further

## Release notes

Notes: no-notes

### CODE REVIEW

File: app/src/main-process/desktop-file-transport.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    const draining: boolean = s.write(chunk, e =>
+      e ? reject(e) : resolve(draining)
+    )
+  })
+
+const getFilePrefix = (d = new Date()) => d.toISOString().split('T', 1)[0]
+const getFilePath = (p: string) => join(p, `${getFilePrefix()}${fileSuffix}`)
+
+const pruneDirectory = async (p: string) => {
+  const treshold = offsetFromNow(-14, 'days')
```
Comment (tidy-dev) [inline app/src/main-process/desktop-file-transport.ts]:
```suggestion
  const threshold = offsetFromNow(-14, 'days')
```

File: app/src/main-process/desktop-file-transport.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  return undefined
+}
+
+/**
+ * A re-implementation of the winston-daily-rotate-file module
+ *
+ * winston-daily-rotate-file depends on moment.js which we're trying to unship
+ * so we're using this instead.
+ *
+ * Please note that this is in no way a general purpose transpot like
```
Comment (sergiou87) [inline app/src/main-process/desktop-file-transport.ts]: Nit
```suggestion
 * Please note that this is in no way a general purpose transport like
```

File: app/src/main-process/desktop-file-transport.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  for (const f of files ?? []) {
+    const m = pathRe.exec(f)
+    const d = m ? Date.parse(m[1]) : NaN
+
+    if (!isNaN(d) && d < threshold) {
+      await unlink(join(p, f)).catch(error('unlink'))
+    }
+  }
+}
```
Comment (sergiou87) [inline app/src/main-process/desktop-file-transport.ts]: Unless I'm missing something, I think this introduces an important change in behavior: the previous implementation kept up to 14 log files, no matter how old they were, while this one deletes log files older than 14 days, no matter how many there are. So if I install and open the app today, generate a log file, and then don't open the app in 15 days… the next day I open it, my first log file will be deleted.

Not the end of the world, but I think it's better to have as much info as possible when users report issues. Who knows if there's something valuable in older logs… 😅

File: app/test/unit/desktop-file-transport-test.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    try {
+      expect(await readdir(d)).toBeArrayOfSize(0)
+      for (let i = 1; i < 20; i++) {
+        const date = `${i}`.padStart(2, '0')
+        globalDate.now = () => Date.parse(`2022-03-${date}T10:00:00.000Z`)
+        globalDate.prototype.toISOString = () => `2022-03-${date}T10:00:00.000Z`
+        await info(t, 'heyo')
+      }
+
+      expect(await readdir(d)).toBeArrayOfSize(14)
```
Comment (sergiou87) [inline app/test/unit/desktop-file-transport-test.ts]: Maybe worth also checking that it deletes the older files, and not the newer ones, as that's critical behavior?

**Review by sergiou87 — COMMENTED** [review]
Awesome work!!! I think it all makes sense, and after a thorough review, I don't see any obvious flaws 🤞 

I only noticed a change in behavior that I think is worth reverting (pruning logs older than 14 days vs pruning older logs when the user has more than 14), but other than that… 🤩

File: app/test/unit/desktop-file-transport-test.ts (hunk 4 above)
Comment (niik) [reply app/test/unit/desktop-file-transport-test.ts]: Great point!

File: app/src/main-process/desktop-file-transport.ts (hunk 3 above)
Comment (niik) [reply app/src/main-process/desktop-file-transport.ts]: I would argue that retaining based on age is a better approach here, we rarely think of retention in any other metric than time. That said I think it's important to not introduce more change than we have to here given the delicate nature of this logic so I've changed it to retain 14 files, regardless of age.

**Review by sergiou87 — APPROVED** [review]
Looks great! I will play a bit with it tomorrow 😄  But I'll approve it now anyway in case you're in a hurry to merge it.

**Review by sergiou87 — DISMISSED** [review]
Tested and seems to work as expected 🤞  :shipit: ❗

**Review by sergiou87 — APPROVED** [review]
Approving again after resolving conflicts…

**Pull request merged**